pkg/kubelet/pod_workers.go

```diff
@@ -734,10 +748,71 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
 	return gracePeriod, status.gracePeriod != 0 && status.gracePeriod != gracePeriod
 }
 
+// allowPodStart tries to start the pod and returns true if allowed, otherwise
+// it requeues and returns false.
+func (p *podWorkers) allowPodStart(pod *v1.Pod) bool {
+	if !kubetypes.IsStaticPod(pod) {
+		return true
+	}
+	p.podLock.Lock()
+	defer p.podLock.Unlock()
+	if !p.isStaticPodAllowedToStart(pod) {
+		p.workQueue.Enqueue(pod.UID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
+		p.podSyncStatuses[pod.UID].working = false
+		return false
+	}
+	p.allowNextStaticPod(kubecontainer.GetPodFullName(pod))
+	return true
+}
+
+// isStaticPodAllowedToStart returns true if there is no other started static
+// pod with the same full name and pod is the oldest static pod waiting to
+// start.
+func (p *podWorkers) isStaticPodAllowedToStart(pod *v1.Pod) bool {
+	fullname := kubecontainer.GetPodFullName(pod)
+	uid, started := p.startedStaticPodsByFullname[fullname]
+	if started {
+		if uid == pod.UID {
+			// pod is already started
+			return true
+		}
+		return false
+	}
+
+	waitingToStartPods := p.waitingToStartStaticPodsByFullname[fullname]
+	if len(waitingToStartPods) == 0 {
+		return false
```

Comment:
If we are not waiting, shouldn't this be true?  "A static pod is allowed to start if there are no waiting static pods with that full name"?